### PR TITLE
Form UWUW material name

### PR DESCRIPTION
**Pull request by bam241**
I figured that as we are slowly moving on with the MaterialLibrary C++ API at some point we would need something like that...
this allows to form a UWUW material name, including the density if provided.

this solves #1188

### CODE REVIEW

File: pyne/material.pyx [hunk 1]
```diff
@@ -345,6 +345,14 @@ cdef class _Material:
         cdef std_string card
         card = self.mat_pointer.mcnp(frac_type.encode())
         return card.decode()
+    
+    def mcnp(self):
```
Comment (pshriwise) [inline pyne/material.pyx]: The method name here is incorrect.
```suggestion
    def uwuw(self):
```

File: pyne/material.pyx [hunk 2]
```diff
@@ -345,6 +345,14 @@ cdef class _Material:
...
+        """mcnp(frac_type)
```
Comment (pshriwise) [inline pyne/material.pyx]: Docstring should be updated as well. There is no `frac_type` argument and function name needs updating, same as above.

File: src/material.cpp [hunk 3]
```diff
@@ -594,6 +594,26 @@ std::string pyne::Material::openmc(std::string frac_type) {
   return oss.str();
 }
 
+///---------------------------------------------------------------------------//
+std::string pyne::Material::uwuw() {
+  // standard uwuw material name is : "mat:<Name of Material>/rho:<density>" 
+  if (! metadata.isMember("name")) {
+    pyne::warning("The material has no name!!");
```
Comment (pshriwise) [inline src/material.cpp]:
```suggestion
    pyne::warning("The material has no name");
```

File: tests/test_material.py [hunk 4]
```diff
@@ -1290,7 +1290,47 @@ def test_mcnp_mat0():
     assert_equal(mass, mass_exp)
 
 
+def uwuw():
```
Comment (pshriwise) [inline tests/test_material.py]: This test won't be found by `nosetests`. It needs to have a name like `test_*`
```suggestion
def test_uwuw():
```

File: src/material.cpp [hunk 5]
```diff
@@ -594,6 +594,26 @@ std::string pyne::Material::openmc(std::string frac_type) {
...
+  }
+  std::string uwuw_name = "mat:<";
+  uwuw_name += metadata["name"].asString();
+  if (density > 0) {
+    uwuw_name += "/rho:" +std::to_string(density);
+  } else {
+    pyne::warning("No Density defined for this Material");
+  }
+
+  uwuw_name += ">";
```
Comment (pshriwise) [inline src/material.cpp]:
```suggestion
```

File: src/material.cpp [hunk 6]
```diff
@@ -594,6 +594,26 @@ std::string pyne::Material::openmc(std::string frac_type) {
...
 }
 
+///---------------------------------------------------------------------------//
+std::string pyne::Material::uwuw() {
+  // standard uwuw material name is : "mat:<Name of Material>/rho:<density>" 
+  if (! metadata.isMember("name")) {
+    pyne::warning("The material has no name!!");
+    return "";
+  }
+  std::string uwuw_name = "mat:<";
```
Comment (pshriwise) [inline src/material.cpp]:
```suggestion
  std::string uwuw_name = "mat:";
```

**Review by pshriwise — CHANGES_REQUESTED** [review]
Good idea.

The format for names isn't quite right though. UWUW like the leading and trailing `<`/`>` characters.

File: tests/test_material.py [hunk 7]
```diff
@@ -1290,7 +1290,47 @@ def test_mcnp_mat0():
...
+    leu2 = Material(nucvec={'U235': 0.04, 'U238': 0.96},
+                   metadata={'mat_number': 2,
+                          'table_ids': {'92235':'15c', '92238':'25c'},
+                          'mat_name':'LEU',
+                          'source':'Some URL',
+                          'comments': ('this is a long comment that will definitly '
+                                       'go over the 80 character limit, for science'),
+                          'name':'leu'})
+    uwuw_name = leu2.uwuw()
+    name_exp = ('<mat:leu>')
```
Comment (bam241) [inline tests/test_material.py]:
```suggestion
    name_exp = ('mat:leu')
```

File: tests/test_material.py [hunk 8]
```diff
@@ -1290,7 +1290,47 @@ def test_mcnp_mat0():
...
+                          'table_ids': {'92235':'15c', '92238':'25c'},
+                          'mat_name':'LEU',
+                          'source':'Some URL',
+                          'comments': ('this is a long comment that will definitly '
+                                       'go over the 80 character limit, for science'),
+                          'name':'leu'},
+                   density=19.1)
+
+    uwuw_name = leu.uwuw()
+    name_exp = ('<mat:leu/rho:19.1>')
```
Comment (bam241) [inline tests/test_material.py]:
```suggestion
    name_exp = ('mat:leu/rho:19.1')
```

File: pyne/material.pyx [hunk 9]
```diff
@@ -345,6 +345,14 @@ cdef class _Material:
         cdef std_string card
         card = self.mat_pointer.mcnp(frac_type.encode())
         return card.decode()
+    
+    def uwuw(self):
+        """mcnp(frac_type)
```
Comment (bam241) [inline pyne/material.pyx]:
```suggestion
        """uwuw()
```

**Comment by bam241** [comment]
thx for the review @pshriwise and for the suggestion made my work so easy !!

**Review by gonuke — COMMENTED** [review]
My only comment is about the name of the method.  `uwuw` doesn't communicate what it's doing very well.  What about `get_uwuw_name`?

(Note: in general, I have a preference for method names to be verbs, but there are already many that are not, so I can't impose it as a norm)

**Comment by bam241** [comment]
> My only comment is about the name of the method. `uwuw` doesn't communicate what it's doing very well. What about `get_uwuw_name`?
> 
> (Note: in general, I have a preference for method names to be verbs, but there are already many that are not, so I can't impose it as a norm)

I am fine with that suggestion, but does it mean we should eventually change the `mcnp()` and the `fluka() ones ?

File: tests/test_material.py [hunk 10]
```diff
@@ -1290,7 +1290,47 @@ def test_mcnp_mat0():
     assert_equal(mass, mass_exp)
 
 
+def test_uwuw():
+    leu = Material(nucvec={'U235': 0.04, 'U238': 0.96},
+                   metadata={'mat_number': 2,
+                          'table_ids': {'92235':'15c', '92238':'25c'},
+                          'mat_name':'LEU',
```
Comment (pshriwise) [inline tests/test_material.py]: Help me out here. Why both `mat_name` and `name` in the material def?

Comment (bam241) [reply tests/test_material.py]: this is actually a good question, I copy pasted the one form `mcnp()` test... I should probably fix both of them !

Comment (bam241) [reply tests/test_material.py]: looking at the source code of material.cpp, it seems that openmc is using the `mat_name` metadata to name the material where all the others uses `name`.

Here is what I suggest:
In this PR: fix my test about `uwuw` removing mat_name
Issue an issue: to remove mention to mat_name in `openmc()` and in the test_material.py 
@pshriwise what do you thing ?

Shall we also implement something to catch old `mat_name` and convert them into name ? (for backward compatibility)

Comment (pshriwise) [reply tests/test_material.py]: That sounds like a good plan to me!

**Comment by gonuke** [comment]
> > My only comment is about the name of the method. `uwuw` doesn't communicate what it's doing very well. What about `get_uwuw_name`?
> > (Note: in general, I have a preference for method names to be verbs, but there are already many that are not, so I can't impose it as a norm)
> 
> I am fine with that suggestion, but does it mean we should eventually change the `mcnp()` and the `fluka() ones ?

Perhaps - but they are qualitatively different.  This new method extracts a single string out of the metadata in a particular format.  The others are a little more comprehensive (although still not verbs).

**Comment by bam241** [comment]
@pshriwise  ?

**Review by pshriwise — APPROVED** [review]

**Comment by pshriwise** [comment]
Will merge this evening if there are no further comments. Thanks @bam241!

**Comment by gonuke** [comment]
Thanks @bam241

**Pull request merged**